bazaarvoice/emodb
# Megabus Touch functionality of Emo Documents.

## Issue #

Legion team would like to have a feature to reprocess emo documents and be able to instruct EMODB to remit the document onto the Megabus.

## What Are We Doing Here?

Added End points for touching a single documents as well a stream of multiple documents. The implementation is simple which binds the stream to Refs and sends them to Megabus Ref topic, from which the Resolver will pick them up and finally end up in the actual Megabus. 

Current support is using the end points directly. No client has been added, but that can be quickly added along with the relevant Jersey tests if thats the requirement.  

## Risk
very Low. The implementation is independent of the existing Megabus functionality. so don't expect any regressions. 

## How to Test and Verify

> Tested locally:
> svaddi-retina: curl -XPOST localhost:8082/megabus/1/review:homedepot/key1
> {"success":true}
> 
> ./kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic emodb.internal.megabus.ref --from-beginning | jq .
> [
>   {
>     "table": "review:homedepot",
>     "key": "key1",
>     "changeId": "8f16ec20-37d4-11ea-a5e6-0e0d17a0c2c9",
>     "readTime": "2020-01-15T20:21:06.146Z"
>   }
> ]


> sujith.vaddi$ curl -XPOST -H "Content-Type: application/json" -v "http://localhost:8082/megabus/1/_stream" --data-binary '[{"table":"review:walmart", "key":"key2","changeId":"44f344e0-37c5-11ea-9389-0e0d17a0c2c9","readTime":"2020-01-15T18:31:39.310Z"},{"table":"review:testcustomer", "key":"key3","changeId":"44f344e0-37c5-11ea-9389-0e0d17a0c2d5","readTime":"2020-01-15T20:31:39.310Z"}]'
> 
> svaddi-retina:bin sujith.vaddi$ ./kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic emodb.internal.megabus.ref --from-beginning | jq .
> [
>   {
>     "table": "review:testcustomer",
>     "key": "key3",
>     "changeId": "44f344e0-37c5-11ea-9389-0e0d17a0c2d5",
>     "readTime": "2020-01-15T20:31:39.310Z"
>   }
> ]
> [
>   {
>     "table": "review:walmart",
>     "key": "key2",
>     "changeId": "44f344e0-37c5-11ea-9389-0e0d17a0c2c9",
>     "readTime": "2020-01-15T18:31:39.310Z"
>   }
> ]

Made sure we see the document in actual Megabus Topic: In the below example one was when the document is created, and other one was from the TOUCH:

> svaddi-retina:bin sujith.vaddi$ ./kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic emodb.megabus --from-beginning | jq .
> 
> {
>   "~deleted": false,
>   "~firstUpdateAt": "2020-01-15T21:09:30.800Z",
>   "~id": "sujith1",
>   "~lastMutateAt": "2020-01-15T21:09:30.800Z",
>   "~lastUpdateAt": "2020-01-15T21:09:30.800Z",
>   "~signature": "5693f9808826a95c32178ebecd8851b8",
>   "~table": "review:testcustomer",
>   "~version": 1,
>   "author": "Bob",
>   "client": "TestCustomer",
>   "rating": 5,
>   "title": "Best Ever!",
>   "type": "review"
> }
> {
>   "~deleted": false,
>   "~firstUpdateAt": "2020-01-15T21:09:30.800Z",
>   "~id": "sujith1",
>   "~lastMutateAt": "2020-01-15T21:09:30.800Z",
>   "~lastUpdateAt": "2020-01-15T21:09:30.800Z",
>   "~signature": "5693f9808826a95c32178ebecd8851b8",
>   "~table": "review:testcustomer",
>   "~version": 1,
>   "author": "Bob",
>   "client": "TestCustomer",
>   "rating": 5,
>   "title": "Best Ever!",
>   "type": "review"
> }

**LATEST TESTING WORK - API endpoints and data is changed:** 
```
curl -XPOST localhost:8082/megabus/1/_touch/review:sujith2/key1
curl -XPOST -H "Content-Type: application/json" -v "http://localhost:8082/megabus/1/_touch/_stream" --data-binary '[{"~table":"review:walmart", "~id":"key2"},{"~table":"review:testcustomer", "~id":"key3"}]'
```
